framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -382,41 +379,62 @@ public void processClassTree(ClassTree classTree) {
             }
         }
 
+        checkExtendsImplements(classTree);
+
+        super.visitClass(classTree, null);
+    }
+
+    /**
+     * If "@B class Y extends @A X {}", then enforce that @B must be a subtype of @A.
+     *
+     * <p>Also validate the types of the extends and implements clauses.
+     *
+     * @param classTree class tree to check
+     */
+    protected void checkExtendsImplements(ClassTree classTree) {
+        if (TypesUtils.isAnonymous(TreeUtils.typeOf(classTree))) {
+            // Don't check extends clause on anonymous classes.
+            return;
+        }
+
+        AnnotatedTypeMirror classType = atypeFactory.getAnnotatedType(classTree);
```

Comment:
Note that this was already calculated on line 364. If this computation isn't cached, it might be worth passing as a second parameter.

Reply:
It's cached.  But thanks for pointing this out.